### PR TITLE
Feature/exams overview

### PR DESCRIPTION
**Pull request by Djoleoo**

### CODE REVIEW

File: Konteh/back-office-app/src/app/exams-overview/exams-overview.component.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    private router: Router
+  ) {}
+
+  ngOnInit(): void {
+    this.route.queryParams.subscribe(params => {
+      this.searchText = params['searchText'] || '';  
+      this.currentPage = +params['page'] || 0;       
+      this.pageSize = +params['pageSize'] || 5;      
+
+      this.loadExams(this.currentPage, this.pageSize, this.searchText);
```
Comment (MilosNesh) [inline Konteh/back-office-app/src/app/exams-overview/exams-overview.component.ts:32]: I don't think you should send this.current , this.page , this.searchText as parameters

File: Konteh/back-office-app/src/app/exams-overview/exams-overview.component.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    let request = new GetExamsForOverviewQuery({
+      text: searchText,
+      page: this.currentPage,
+      pageSize: this.pageSize
+    });
+    this.examClient.getAllExams(request).subscribe((response: GetExamsForOverviewResponse) => {
+      this.dataSource.data = response.exams || [];
+      this.totalExams = response.length || 0;
+    }, error => {
+      console.error('Error loading exams: ', error);
```
Comment (MilosNesh) [inline Konteh/back-office-app/src/app/exams-overview/exams-overview.component.ts:46]: remove console.log

File: Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        public RequestHandler(IRepository<Exam> examRepository)
+        {
+            _examRepository = examRepository;
+        }
+
+        public async Task<Response> Handle(Query request, CancellationToken cancellationToken)
+        {
+            IQueryable<Exam> query;
+            int length;
+
```
Comment (MilosNesh) [inline Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs:49]: can this be shorter

File: Konteh/Konteh.BackOfficeApi/Features/Exams/ExamController.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+public class ExamController : Controller
+{
+    private readonly IMediator _mediator;
+
+    public ExamController(IMediator mediator)
+    {
+        _mediator = mediator;
+    }
+
+    [HttpPut]
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Exams/ExamController.cs:18]: put???

File: Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    public class Response
+    {
+        public IEnumerable<ExamResponse> Exams { get; set; } = new List<ExamResponse>();
+        public int Length { get; set; }
+    }
+    public class ExamResponse
+    {
+        public int Id { get; set; }
+        public CandidateInformation Candidate { get; set; } = new CandidateInformation();
+        public int TotalNumberOfQuestions { get; set; }
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs:24]: you can combine these two fields into a string I think

File: Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    public class Response
+    {
+        public IEnumerable<ExamResponse> Exams { get; set; } = new List<ExamResponse>();
+        public int Length { get; set; }
+    }
+    public class ExamResponse
+    {
+        public int Id { get; set; }
+        public CandidateInformation Candidate { get; set; } = new CandidateInformation();
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs:23]: again, no need for a separate object here, just return the Candidate's full name

File: Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                query = query
+                    .Skip(request.Page * request.PageSize)
+                    .Take(request.PageSize)
+                    .OrderByDescending(exam => exam.StartTime);
+            }
+            else
+            {
+                query = _examRepository.GetPaged(request.Page, request.PageSize)
+                    .OrderByDescending(exam => exam.StartTime)
+                    .AsQueryable();
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs:69]: Why turn it into a queryable? You already pulled the data into memory. There has be a cleaner solution here.

File: Konteh/back-office-app/src/app/exams-overview/exams-overview.component.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  loadExams(): void {
+    let request = new GetExamsForOverviewQuery({
+      text: this.searchText,
+      page: this.currentPage,
+      pageSize: this.pageSize
+    });
+    this.examClient.getAllExams(request).subscribe((response: GetExamsForOverviewResponse) => {
+      this.dataSource.data = response.exams || [];
+      this.totalExams = response.length || 0;
+    }, error => {
```
Comment (laslorajda) [inline Konteh/back-office-app/src/app/exams-overview/exams-overview.component.ts:45]: if you do nothing in the error block it's useless code. Delete it.

File: Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        {
+            _examRepository = examRepository;
+        }
+
+        public Task<Response> Handle(Query request, CancellationToken cancellationToken)
+        {
+            string searchText = "";
+            if (request.Text != null)
+            {
+                searchText = request.Text.Trim().ToLower();
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs:42]: I'm fine with Trim, ToLower does nothing except create another string

File: Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+                exam.Candidate.Name.Contains(searchText) ||
+                exam.Candidate.Surname.Contains(searchText));
+
+            var length = query.Count();
+
+            var exams = query
+                .OrderByDescending(exam => exam.StartTime)
+                .Skip(request.Page * request.PageSize)
+                .Take(request.PageSize)
+                .ToList();
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs:55]: ToListAsync please

File: Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+            return Task.FromResult(response);
+        }
+        private int CountRightAnswers(Exam exam)
+        {
+            var rightAnswerCount = 0;
+
+            foreach (var eq in exam.ExamQuestions)
+            {
+                var rightAnswers = eq.Question.Answers.Where(answer => answer.IsCorrect).ToList();
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs:80]: I think determining whether an ExamQuestion is answered correctly or not is the responsibility of the ExamQuestion itself. That way this business logic is not hidden in some random getter + writing unit tests(which I don't see) will be easier

File: Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+                .OrderByDescending(exam => exam.StartTime)
+                .Skip(request.Page * request.PageSize)
+                .Take(request.PageSize)
+                .ToList();
+
+            var examResult = exams.Select(exam => new ExamResponse
+            {
+                Id = exam.Id,
+                CandidateName = $"{exam.Candidate.Name} {exam.Candidate.Surname}",
+                Score = $"{CountRightAnswers(exam)}/{exam.ExamQuestions.Count}",
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Exams/GetExamsForOverview.cs:61]: And once you move that logic to the exam question, you can just to exam.examquestions.count(x => x.iscorrect) or whatever

File: Konteh/Konteh.Domain/ExamStatus.cs [hunk 13]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Konteh.Domain;
+
+public enum ExamStatus
+{
+    InProgress,
```
Comment (laslorajda) [inline Konteh/Konteh.Domain/ExamStatus.cs:5]: please give them values

**Review by zelick — APPROVED** [review]
perfecto

**Pull request merged**